### ISSUE TITLE
Consider excluding script_template files from license-maven-plugin

### ISSUE THREAD

**Issue by imagejan** [scijava/pom-scijava-base#4]
In order to avoid adding license headers to script templates, we should maybe add something like:

```
<excludes>
  <exclude>/src/main/resources/script_templates/*</exclude>
</excludes>
```

to here:
https://github.com/scijava/pom-scijava-base/blob/46e2db82d2732748f9550f552d63dd11af972ec2/pom.xml#L527-L546

I didn't test yet if this actually works.

**Comment by ctrueden** [comment]
Great idea. I tested this awhile back, and couldn't make it work. I tested it again today, and still couldn't make it work. I tried things like:

* `/src/main/resources/script_templates/*`
* `/src/main/resources/script_templates/.*`
* `/src/main/resources/script_templates/**`
* `src/main/resources/script_templates/*`
* `src/main/resources/script_templates/.*`
* `src/main/resources/script_templates/**`
* `**/script_templates/**`

Next step is to run with `mvn -X` and read the output carefully, to better understand how the exclusion matching works. I ran out of time this morning, but will try again tonight or tomorrow.

**Comment by ctrueden** [comment]
The following patterns do work:

* `main/resources/script_templates/**`
* `/main/resources/script_templates/**`
* `/main/resources/script_templates/`
* `**/main/resources/script_templates/`
* `%regex[.*main/resources/script_templates/.*]`
* `%regex[.*/script_templates/.*]`

And the following patterns do NOT work:

* `src/main/resources/script_templates/**` (!!)
* `resources/script_templates/`
* `/resources/script_templates/`
* `%regex[.*/main/resources/script_templates/.*]`
* `%regex[/resources/script_templates/.*]`

Conclusions:

* The `**` suffix is not strictly necessary; the license-maven-plugin appends `**` whenever the pattern ends with `/` (I checked the source code). But it is clearer to include it, IMHO.
* The `%regex[...]` style appears to require a match over the entire pattern, necessitating leading and trailing `.*` in common cases. We don't need regex here.
* Somehow, it appears that the `src` directory is not included in the path being matched against? This might be a bug; for maximum future-proofing, we can use a leading `**` glob as well as trailing such glob.

So for now, I am going with `**/script_templates/**` as a working pattern that is hopefully sufficiently general without excluding any desired cases. I have included this directly in the imagej-legacy POM (imagej/imagej-legacy@b7b0d6e5f26de6f54008853df07db85ebda9d984) rather than hardcoding here in pom-scijava-base, because some projects might want license headers included in their script templates; I don't know.

**Comment by ctrueden** [comment]
I also pushed an analogous exclusion for `script_templates` for the following projects:

* fiji/Colocalisation_Analysis
* fiji/KymographBuilder
* imagej/imagej-itk
* imagej/imagej-matlab
* imagej/imagej-scripting
* scijava/scripting-beanshell
* scijava/scripting-groovy
* scijava/scripting-javascript
* scijava/scripting-jruby
* scijava/scripting-jython
* scijava/scripting-matlab
* scijava/scripting-renjin
* scijava/scripting-scala

**Comment by aherbert** [comment]
Note:

Posted in the event this may be useful to anyone wanting to exclude more files from the license maven plugin in child projects of scijava.

The use of the `<license.excludes>` property in `pom-scijava-base` requires a specific way to override the excludes for child projects. I discovered this when updating a project that was created prior to the property changes in `pom-scijava-base`.

A project inheriting from  from `pom-scijava-base` (including `pom-scijava`) that wants to exclude other files from the license-maven-plugin cannot use the `<configuration><excludes>` property with `<exclude>` elements as given in the example [update-file-header](https://www.mojohaus.org/license-maven-plugin/examples/update-file-header-config.html) documentation for the license-maven-plugin. Child projects must directly replace the single string set in the `<license.excludes>` property by an optionally comma-delimited single string.

This will not work:
```xml
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>license-maven-plugin</artifactId>
        <configuration>
          <excludes>
            <exclude>**/*.json</exclude>
            <exclude>**/*.xml</exclude>
          </excludes>
        </configuration>
      </plugin>
```
This will:
```xml
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>license-maven-plugin</artifactId>
        <configuration>
          <excludes>**/*.json,**/*.xml</excludes>
        </configuration>
      </plugin>
```
Or:
```xml
<properties>
   <license.excludes>**/*.json,**/*.xml</license.excludes>
</properties>
```

Running the plugin with the maven `-X` flag allows full debugging of the configuration XML that maven extracts for the license-maven-plugin and the resulting configuration of the license-maven-plugin mojo. An extract of the debug XML for the non-functional example above is:
```xml
<excludes>
    <exclude>**/*.json.xml</exclude>
    <exclude>**/*.xml</exclude>${license.excludes}</excludes>
```
I presume when the `${license.excludes}` property is not empty then maven rejects the mixture of plain text and the `<exclude>` elements. Options would be: (1) to ignore the non-empty `${license.excludes}` property; or (2) ignore the non-empty `<excludes>` configuration that contains child elements. In this case maven is choosing option 2. If the `<excludes>` configuration does not contain child elements then maven chooses option 1.

This is all performed automatically by maven based on the following mojo code annotated with `org.apache.maven.plugins.annotations.Parameter`:
```java
    @Parameter( property = "license.excludes" )
    private String[] excludes;
```
The behaviour would be the same for any array-based mojo parameter that has a property override.

**Comment by ctrueden** [comment]
@aherbert Thanks for the note. I agree that getting the license-maven-plugin's exclusion rules right is very tricky. For one thing, the docs talk about "regexp", but it doesn't appear to use regex, but rather globs. And it seems very finicky what works and what doesn't. Adding on top of that, the issue you describe with array parameters, and it's a recipe for frustration.

To help ease this a little, I made the following change: 473e997ed2e5cd5096d717a5274f90ff70116ead

This introduces a new property `scijava.excludedLicensePatterns` with the default exclusion values (currently `**/script_templates/**`). So then in your POM, you can put the following line in your `<properties>`:
```xml
<license.excludes>${scijava.excludedLicensePatterns},**/*.json,**/*.xml</license.excludes>
```
To extend the patterns rather than replacing them. Might still be tricky to get the patterns themselves right, but at least they'll get appended properly.

You might also want to consider filing an issue (or PR) upstream (https://github.com/mojohaus/license-maven-plugin) about the documentation errors/inaccuracies. The maintainer is not lightning fast, but I have successfully gotten changes merged upstream to that project in the past.

**Comment by aherbert** [comment]
I agree with the issue with regexp. When searching to find out what was wrong with my expressions I traced the use of the excludes String[]. It is passed to an [org.codehaus.plexus.util.DirectoryScanner](https://codehaus-plexus.github.io/plexus-utils/apidocs/org/codehaus/plexus/util/DirectoryScanner.html).

This has all the documentation on what regexp to use. Reading that it seems to agree with all your experimentation listed above, i.e. use `**` for generic paths, `*` for wildcards, no support for `%regex[...]`. In particular it has notes on the behaviour for when the path starts with `/` or not. It may also have different behaviour for windows and linux with regard to the file separator but I've only tested on a linux platform.

With regard to the search defaults for the `license-maven-plugin`, the default roots are set to use `"src", "target/generated-sources", "target/processed-sources"`. So `src` should be included but given you had issues matching it this may be due to the presence/absence of a leading `/` from either your regexp or the path generated by the DirectoryScanner starting from `src`; or because the path starts below `src`, e.g. with `main/java/...`.

I'll update my pom to use `<license.excludes>` as suggested.
